docs/proposals/20220603-kubernetes-vendor.md

```diff
@@ -0,0 +1,235 @@
+
+We can provide native support for ease of use.
+
+## Goals
+
+- Build the envd environments with the buildkitd pod on Kubernetes
+- Run environments on Kubernetes
+  - Compose Kubernetes resources (e.g. deployments, services, pods) when run `envd up`.
+  - Push the image to a registry.
+  - Port forward for sshd and jupyter notebook service.
+  - Sync the pod file system with the local host file system.
+- Manage environments on Kubernetes
+  - List environments running on Kubernetes, add a new field `vendor` when run `envd get envs`
+  - List dependencies of the environment on Kubernetes
+
+## Non-Goals
+
+- Monitoring on Kubernetes
+- Pause and unpause environments on Kubernetes (Kubernetes does not support the primitives)
+- Manage envd images on Kubernetes
+
+## Proposal
+
+### User stories
+
+#### CI/CD with envd
+
+As a ML infra engineer, I want to build the images with CI/CD tools. Docker may not be set up in the CI/CD runners, so that envd should support to build with buildkitd running on the Kubernetes.
+
+#### Remote development
+
+As a AI/ML engineer, I want to develop on the remote cluster (managed by Kubernetes).
+
+### CLI behavior
+
+We need to support primitives like `run`:
+
+```
+$ envd run --env my-env --image my-image
+```
+
+User may use `envd` to build the image, and use it on Kubernetes. Thus they need to push the image to a registry. `envd up` may not satisfy the requirements.
+
+The end-to-end process will be:
+
+```
+$ envd context create --name test --builder-name test --use --builder kube-pod --runner kubernetes --runner-config kube-pod-config.yaml
+$ envd build
+$ envd push
+$ envd run --env test --image test
+
+or 
+
+$ envd context create --name test --builder-name test --use --builder kube-pod --runner kubernetes --runner-config kube-pod-config.yaml
+$ envd up
+```
+
+### Implementation details/Notes/Constraints
+
+![](../images/proposals/k8s.png)
+
+#### envd contexts
+
+envd uses buildkitd container in the local docker host to build the images by default. `context` command will be introduced to support remote build.
+
+```
+NAME:
+   envd context - Manage envd contexts
+
+USAGE:
+   envd context command [command options] [arguments...]
+
+COMMANDS:
+   create   Create envd context
+   ls       List envd contexts
+   rm       Remove envd context
+   use      Use the specified envd context
+   help, h  Shows a list of commands or help for one command
+
+OPTIONS:
+   --help, -h  show help (default: false)
+```
+
+The structure of the envd context is:
+
+```go
+type Context struct {
+	Name          string      `json:"name,omitempty"`
+	Builder       BuilderType `json:"builder,omitempty"`
+	BuilderSocket string      `json:"builder_socket,omitempty"`
```

Comment:
How about `BuilderAddr`, which could accept:

- `unix:///var/run/docker.sock`
- `tcp://192.168.1.108:7890` (if there is a buildkit exposed)

Reply:
LGTM